edualert/common/middleware.py

```diff
@@ -87,3 +90,88 @@ def process_response(self, request, response):
         print("{}{}[MIDDLWARE OVERHEAD: {} seconds]{}\n".format(*replace_tuple))
 
         return response
+
+
+class RequestActivityTrackerMiddleware:
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        # read body before it's read by `rest_framework.Request` so that it's saved on the `django.http.HttpRequest`
+        # object, this will allows us to read it again after the response is retrieved
+        _ = request.body
+
+        # pass the request down the middleware chain
+        response = self.get_response(request)
+
+        # Filter request by method
+        if request.method in ['OPTIONS', 'HEAD', 'GET']:
+            return response
+
+        # Only track requests for identifiable users
+        user = request.user
+        if not user or user.is_anonymous or not user.id:
+            return response
+
+        # Save request information in cache to be extracted and saved to a persistent storage
+        cache_value_raw = {
+            'timestamp_ms': int(time.time() * 1000),
+            'user_id': request.user.id,
+            'method': request.method,
+            'path': request.get_full_path_info(),
+            'status_code': response.status_code,
+            'request_body': _get_request_body(request),
+        }
+        cache_key = 'track_request_{}'.format(uuid.uuid4())
+        cache_value = json.dumps(cache_value_raw)
+        cache.set(cache_key, cache_value, timeout=None)
+
+        return response
+
+
+def _get_request_body(request):
+    """
+    Return a json object when content type is 'application/json'. Sensitive fields in the object are replaced.
+    """
+    request_body = None
+
+    if request.content_type:
+        # return masked content if it's json
+        if request.content_type == 'application/json':
```

Comment:
can't we have a single if with 2 conditions?

Reply:
I updated the PR and merged the two `if` statements. Initially I wasn't sure if JSON was the only content type that needed handling. If you think we'd need to handle something other than JSON as well please let me know.